jyotisha/panchaanga/temporal/festival/applier/solar.py

```diff
@@ -22,6 +31,73 @@ def assign_all(self):
+
+    for d in range(self.panchaanga.duration_prior_padding, self.panchaanga.duration + 1):
+      if self.daily_panchaangas[d].solar_sidereal_date_sunset.month_transition is not None:
+        punya_kaala_str = SANKRANTI_PUNYAKALA_NAMES[self.daily_panchaangas[d + 1].solar_sidereal_date_sunset.month] + '-puNyakAlaH'
+        jd_transition = self.daily_panchaangas[d].solar_sidereal_date_sunset.month_transition
+        # TODO: convert carefully to relative nadikas!
+        punya_kaala_start_jd = jd_transition - PUNYA_KAALA[self.daily_panchaangas[d + 1].solar_sidereal_date_sunset.month][0] * 1/60
+        punya_kaala_end_jd = jd_transition + PUNYA_KAALA[self.daily_panchaangas[d + 1].solar_sidereal_date_sunset.month][1] * 1/60
+        if punya_kaala_start_jd < self.daily_panchaangas[d].jd_sunrise:
+          fday = d - 1
+        else:
+          fday = d
+        # self.panchaanga.add_festival_instance(festival_instance=FestivalInstance(name=punya_kaala_str, interval=Interval(jd_start=punya_kaala_start_jd, jd_end=punya_kaala_end_jd)), date=self.daily_panchaangas[fday].date)
+        self.daily_panchaangas[fday].festival_id_to_instance[punya_kaala_str] = (FestivalInstance(name=punya_kaala_str, interval=Interval(jd_start=punya_kaala_start_jd, jd_end=punya_kaala_end_jd)))
+
+      if self.daily_panchaangas[d].tropical_date_sunset.month_transition is not None:
+        # Add punyakala
+        punya_kaala_str = TROPICAL_SANKRANTI_PUNYAKALA_NAMES[self.daily_panchaangas[d + 1].tropical_date_sunset.month] + '-puNyakAlaH'
+        jd_transition = self.daily_panchaangas[d].tropical_date_sunset.month_transition
+        # TODO: convert carefully to relative nadikas!
+        punya_kaala_start_jd = jd_transition - PUNYA_KAALA[self.daily_panchaangas[d + 1].tropical_date_sunset.month][0] * 1/60
+        punya_kaala_end_jd = jd_transition + PUNYA_KAALA[self.daily_panchaangas[d + 1].tropical_date_sunset.month][1] * 1/60
+        if punya_kaala_start_jd < self.daily_panchaangas[d].jd_sunrise:
+          fday = d - 1
+        else:
+          fday = d
+        # self.panchaanga.add_festival_instance(festival_instance=FestivalInstance(name=punya_kaala_str, interval=Interval(jd_start=punya_kaala_start_jd, jd_end=punya_kaala_end_jd)), date=self.daily_panchaangas[fday].date)
+        self.daily_panchaangas[fday].festival_id_to_instance[punya_kaala_str] = (FestivalInstance(name=punya_kaala_str, interval=Interval(jd_start=punya_kaala_start_jd, jd_end=punya_kaala_end_jd)))
+
+        # Add tropical sankranti
+        masa_name = RTU_MASA_NAMES[self.daily_panchaangas[d + 1].tropical_date_sunset.month]
+        if jd_transition < self.daily_panchaangas[d].jd_sunrise:
+          fday = d - 1
+        else:
+          fday = d
+        # self.panchaanga.add_festival_instance(festival_instance=FestivalInstance(name=masa_name, interval=Interval(jd_start=None, jd_end=jd_transition)), date=self.daily_panchaangas[fday].date)
```

Comment:
Why not using add_festival_instance?

Reply:
It seems to leave out the time. I thought I noted that somewhere in discussion. Is the commented statement correct?